File: grida-domains/domain-rds/src/main/java/org/grida/entity/base/BaseEntity.java
package org.grida.entity.base;

import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;

import javax.persistence.*;
import java.time.LocalDateTime;

@Getter
@MappedSuperclass
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public abstract class BaseEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    protected Long id;

    @Column
    @CreatedDate
    protected LocalDateTime createdAt;

    @Column
    @LastModifiedDate
    protected LocalDateTime lastModifiedAt;

    protected BaseEntity(LocalDateTime createdAt, LocalDateTime lastModifiedAt) {
        this.createdAt = createdAt;
        this.lastModifiedAt = lastModifiedAt;
    }
}

File: grida-storage/rds-storage/src/main/kotlin/org/grida/persistence/base/BaseEntity.kt
package org.grida.persistence.base

import jakarta.persistence.Column
import jakarta.persistence.EntityListeners
import jakarta.persistence.MappedSuperclass
import org.grida.domain.base.Timestamp
import org.springframework.data.annotation.CreatedDate
import org.springframework.data.annotation.LastModifiedDate
import org.springframework.data.jpa.domain.support.AuditingEntityListener
import java.time.LocalDateTime

@MappedSuperclass
@EntityListeners(AuditingEntityListener::class)
abstract class BaseEntity {

    @CreatedDate
    @Column(updatable = false)
    var createdAt: LocalDateTime = LocalDateTime.MIN

    @LastModifiedDate
    var lastModifiedAt: LocalDateTime = LocalDateTime.MIN

    fun toTimeStamp(): Timestamp {
        return Timestamp(createdAt, lastModifiedAt)
    }
}
